### PR TITLE
Feature: Warn when attempting to send tx with data to non-contract

### LINKED ISSUE #1789: You can lose money by accidentally calling contract functions on an address with no contract.
```
var contract = web3.eth.contract(abi).at(address_you_or_your_app_thinks_has_contract_but_actually_doesn't);
contract.function_that_costs_gas_to_execute(... // wastes money
```

### LINKED ISSUE #4525: eth.getCode returns incorrect values for empty addresses
The injected `web3` instance retuns `0x` when one calls `web3.eth.getCode(address)` when `address` has no contract deployed to it. The normal `web3` however returns `0x0`. This inconsistency is not only annoying but very dangerous. See also https://github.com/Neufund/TypeChain/issues/50 and https://github.com/MetaMask/metamask-extension/issues/1789. I think this issue should be addressed quickly and is a low hanging fruit. Happy to write a test and possibly fix for this but I am not sure where this is actually generated in the code.

### PR DESCRIPTION
**Pull request by HowardBraham**
Fixes #1789
Fixes #4525

I think this is technically an issue between Infura and the ethjs library.  When you call `getCode()` on an address that does not have a contract, Ganache will return `0x0` and Infura will return `0x`.  The ethjs library correctly parses the `0x0`, but does not know how to deal with the `0x`.  But absent fixes from Infura and ethjs, it is now fixed in MetaMask.

I have also improved the error message, which used to say
> ALERT: Transaction Error. Exception thrown in contract code.

And now says the clearer
> ALERT: Trying to call a contract function at an address that is not a contract address.

(This does have to be translated to additional languages.)

I have done regression tests against the following 12 test cases in the table below:

|                                       |                                  Infura                                 |                          Ganache                         |
|---------------------------------------|:-----------------------------------------------------------------------:|:--------------------------------------------------------:|
| call function on nonexistent contract |  **4.9.3** will let you, **PR ???** will stop you  (ALERT was improved) |             will stop you (ALERT was improved)           |
| call function on real contract        |                                  works                                  |                           works                          |
| call function on ERC-20 contract      |                     works and shows as a *TRANSFER*                     |              works and shows as a *TRANSFER*             |
| send ETH to personal address          | works and shows as *CONFIRM*                                            | works and shows as *CONFIRM*                             |
| send ETH to contract without payable  | will stop you (ALERT should be improved)                                | will stop you (ALERT should be improved)                 |
| send ETH to contract with payable     | works (but should possibly show a warning in the future)                | works (but should possibly show a warning in the future) |

Two opportunities for additional improvement (as mentioned in the table above) are the alert when you send ETH to a contract without payable, and a possible warning when you send ETH to a contract with payable (as this is a special case with pitfalls the user should be aware of).

Please let me know if I overlooked any test cases.

### CODE REVIEW

**Comment by frankiebee** [comment]
Infura dosent manipulate responses for getCode so most likely is coming directly from the client code i just want to verify this behavior and will come back to this PR after that.

**Review by frankiebee — DISMISSED** [review]
This is not a bug with metamask/infura but a bug with ganache and ethjs `0x` is `nil` and `0x0` or `0x00` as it should be is 0 my recommendation is to fix the bug directly in ganache and have ethjs handle `0x` as `null`

**Comment by HowardBraham** [comment]
What's the standard?  Returning `0x`, or returning `0x0`?

**Comment by frankiebee** [comment]
`0x` is standard

**Comment by HowardBraham** [comment]
Ahah, okay.  I previously thought `0x0` was the standard.  So it's not an Infura bug.  As you said, it's a bug between Ganache and ethjs.  Is it possible to make that error message clearer from inside ethjs?

**Comment by HowardBraham** [comment]
I found this issue mentioned at Ganache at trufflesuite/ganache-core#51, and interestingly, they claimed to have closed the issue on April 12.

I submitted a new pull request to Ganache at trufflesuite/ganache-core#171 that I believe should actually close that issue.

Does anyone happen to know which package in ethjs contains the relevant `getCode`?  I'm having a little trouble locating it.

**Comment by frankiebee** [comment]
@HowardBraham your looking at https://github.com/ethjs ?

**Comment by HowardBraham** [comment]
@frankiebee Yes, exactly, MetaMask has a dependency on ethjs, ethjs-contract, ethjs-ens, and ethjs-query.  Do we know where this problem might be located?

**Comment by HowardBraham** [comment]
@frankiebee Actually, I have a feeling it might be in https://github.com/ethjs/ethjs-format.  There are probably hundreds of projects that depend on this code, so fixing it is probably not that simple.  I'm going to open an issue over there.

**Comment by HowardBraham** [comment]
@frankiebee After further investigation, I'm not so sure anymore that there's a problem with ethjs.  I think the `0x0` thing was actually a bit of a red herring.  I think the problem is this:

If you POST an `eth_estimateGas` request to Ganache, and the contract does not exist, Ganache will return
`{"id":8379431946159,"jsonrpc":"2.0","error":{"message":"Attempting to run transaction which calls a contract function, but recipient address 0x000000000000000000000000000000000000dead is not a contract address","code":-32000,...`

This will come back as an error in MetaMask.

On the other hand, when you try `eth_estimateGas` with Infura, it does not return an error.  Instead, it returns:
`{"jsonrpc":"2.0","id":8379431946155,"result":"0x53d8"}`

So ***we can't always trust*** the remote node to give an accurate `eth_estimateGas` answer.

Instead, we must:
* Proactively call `eth_getCode`
* If the `txParams` have data AND the code is either falsy, `0x`, or `0x0`
    * Then we're trying to call a contract function on a contract that does not exist, and we should warn the user.
    * Don't even bother trying `eth_estimateGas`.

My pull request already does this, and thus fixes the issues.

**Comment by HowardBraham** [comment]
@frankiebee: ganache-core merged my PR trufflesuite/ganache-core#171

So now future versions of Ganache will no longer return the non-standard `0x0`.

But I think we should keep the functionality in MetaMask to parse `0x0` for maximum compatibility with all RPCs.  Plus, the rest of my changes are necessary for connecting to Infura, which will not give an accurate `eth_estimateGas` answer for no contract.

**Comment by HowardBraham** [comment]
***UPDATE***: This was blocked by #5458, but fixed by #5475.  I have now rebased this PR to the head of `develop`.

**Comment by kumavis** [comment]
@HowardBraham can you expand on why you think sending eth to a payable contract requires a warning?

**Comment by kumavis** [comment]
Perhaps we can improve the language here to something less technical users can understand:
> "Trying to call a contract function at an address that is not a contract address."

File: app/scripts/controllers/transactions/tx-gas-utils.js [hunk 1]
```diff
@@ -56,24 +59,37 @@ class TxGasUtil {
...
+    if (hasRecipient) {
+      const code = await this.query.getCode(recipient)
+      // For an address with no code, geth will return '0x', and ganache-core v2.2.1 will return '0x0'
+      const codeIsEmpty = !code || code === '0x' || code === '0x0'
+      
+      if (codeIsEmpty) {
+        // if there's data in the params, but there's no contract code, it's not a valid transaction
+        if (txParams.data) {
+          const err = new Error()
+          err.errorKey = TRANSACTION_NO_CONTRACT_ERROR_KEY
```
[the remark this answers is not in the document]
Comment (HowardBraham) [reply app/scripts/controllers/transactions/tx-gas-utils.js]: The errorKey is converted to `'transactionErrorNoContract'` in `ui\app\constants\error-keys.js`, and then `"Trying to call a contract function at an address that is not a contract address."` in `app\_locales\en\messages.json`.

**Comment by HowardBraham** [comment]
@kumavis
A proposal for less technical language:
> ALERT: You're trying to call a function that does not exist. This will fail and waste your gas fee.

When sending ETH to a payable contract, I think there should at least be an indication that it's a contract and not an individual. Just a little reminder for the user to think twice and make sure they're doing what they expect. Currently, there's no indication whatsoever.

**Comment by HowardBraham** [comment]
I don't know if you were able to test this, but the correct error message _does_ show up in the UI.

Would a unit test for this actually connect to Infura or geth, or just simulate the connection?  Simply connecting to the internal Ganache might not be enough, because Ganache did not exhibit this bug.  Can someone show me an example of a similar unit test?

**Comment by HowardBraham** [comment]
@kumavis, can you help me figure out how to finish this PR and get it merged?  I answered your questions and asked one above (about unit test examples).  Thanks!

**Comment by kumavis** [comment]
@HowardBraham i guess my primary concern at this point is the creation of an error without a message (even though the message gets set via the localization). @bitpshr @whymarrh do we do this anywhere else?

**Comment by HowardBraham** [comment]
@kumavis @bitpshr @whymarrh I based it on this code, which I know is not exactly the same, because it's not throwing an Error object:

https://github.com/MetaMask/metamask-extension/blob/600f755dbf8d4cfdc152e3d521b537ee9a046a35/ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js#L124-L129

If there's a better pattern to follow, in order to throw an error with a localized message, I'm happy to change it.

**Comment by HowardBraham** [comment]
@kumavis @bitpshr @whymarrh I based it on this code, which I know is not exactly the same, because it's not throwing an Error object:

https://github.com/MetaMask/metamask-extension/blob/600f755dbf8d4cfdc152e3d521b537ee9a046a35/ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js#L124-L129

If there's a better pattern to follow, in order to throw an error with a localized message, I'm happy to change it.

**Comment by kumavis** [comment]
Thanks again @HowardBraham 
made some changes and published a new PR
https://github.com/MetaMask/metamask-extension/pull/5283

**Comment by liliya1987** [comment]
🚖🚏🚗🚖🚝🚖🚐🚘🚘🚘🚘🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚞🚞
🚜🚜🚜🚜🚛🚜🚑🚘🚟🚘🚘🚘🚞🚞🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚐🚗🚗🚘🚗🚗🚐🚐🚐🚐🚐🚐🚐🚐🚐🚐🚐🚐🚐🚐🚐🚐🚐🚐🚘🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚗🚘🚘🚘🚘🚘🚞🚟🚟🚟🚘🚘🚗🚘🚘🚘🚗🚗🚑🚑🚑🚑🚍🚌🚲🚁🚁🚁🚁🚴🚥🚥🚁🚁🚁🚁🚁🚁🚁🚁🚁🚁🚼🚯🌝🌚🌇🍋🍉🌰🌱🍑🍇🍆🍅🍄🍋🍊🍏🍈🍏🌲🌳🌴🍒🍓🍍🌷🍍🍓🍅🍕🍝🍟🍱🍥🍢🍡🍘🍠🍎🍌🍲🍣🍗🍖🍕🍝🍟🍱🍞🍜🍘🍙🍚🍛🍤🍣🍲🍲🍎🍎🍌🐊🐊🐙🐝🐡🐫🐜🐛🐠🐟🐂🐺🐘🐧🐩📷📮✉🎊🎎hhhujopoo

ср, 10 окт. 2018, 14:26 kumavis <notifications@github.com>:

**Pull request merged**